plugin-server/src/kafka/producer.ts

```diff
@@ -1,162 +1,225 @@
+                // If we get a retriable error, bubble that up so that the
+                // caller can retry.
+                throw new DependencyUnavailableError(error.message, 'Kafka', error)
+            } else if ((error as LibrdKafkaError).code === 10) {
+                throw new MessageSizeTooLarge(error.message, error)
             }
-        })
-    )
+
+            throw error
+        }
+    }
+
+    /**
+     * Currently this produces messages in parallel.
+     * If ordering is required then you should use the `produce` method instead in an awaited loop.
+     */
+    async queueMessages(topicMessages: TopicMessage | TopicMessage[]): Promise<void> {
+        topicMessages = Array.isArray(topicMessages) ? topicMessages : [topicMessages]
+
+        await Promise.all(
+            topicMessages.map((record) => {
+                return Promise.all(
+                    record.messages.map((message) =>
+                        this.produce({
+                            topic: record.topic,
+                            key: message.key ? Buffer.from(message.key) : null,
+                            value: message.value ? Buffer.from(message.value) : null,
+                        })
+                    )
+                )
+            })
+        )
+    }
+
+    public async flush() {
+        status.debug('📤', 'flushing_producer')
+
+        return await new Promise((resolve, reject) =>
+            this.producer.flush(10000, (error) => {
+                status.debug('📤', 'flushed_producer')
+                if (error) {
+                    reject(error)
+                } else {
+                    resolve(null)
+                }
+            })
+        )
+    }
+
+    public async disconnect(): Promise<void> {
+        await this.flush()
+
+        status.info('🔌', 'Disconnecting producer')
+        await new Promise<ClientMetrics>((resolve, reject) =>
+            this.producer.disconnect((error: any, data: ClientMetrics) => {
+                status.info('🔌', 'Disconnected producer')
+                if (error) {
+                    reject(error)
+                } else {
+                    resolve(data)
+                }
+            })
+        )
+    }
 }
+
+export const kafkaProducerMessagesQueuedCounter = new Counter({
+    name: 'kafka_producer_messages_queued_total',
+    help: 'Count of messages queued to the Kafka producer, by destination topic.',
+    labelNames: ['topic_name'],
+})
+
+export const kafkaProducerMessagesWrittenCounter = new Counter({
+    name: 'kafka_producer_messages_written_total',
+    help: 'Count of messages written to Kafka, by destination topic.',
+    labelNames: ['topic_name'],
+})
+
+export const kafkaProducerMessagesFailedCounter = new Counter({
+    name: 'kafka_producer_messages_failed_total',
+    help: 'Count of write failures by the Kafka producer, by destination topic.',
+    labelNames: ['topic_name'],
+})
+
+export const ingestEventKafkaProduceLatency = new Summary({
```

Comment:
note: This metric would be better as a histogram.

Reply:
yeah indeed. There is a bigger PR that will be the basis for improvements here. Again I really don't want to change anything here - just improve the code